csrc/scheduler/matmul_utils.cpp

```diff
@@ -314,144 +314,302 @@ std::string isMatmulFusionDefinitionSupported(
-  NVF_ERROR(d_it != roles_map.end(), "Could not find any output D tensors");
-  for (TensorView* tv : d_it->second) {
-    const int64_t N =
-        runtime_info.expressionEvaluator()
-            .evaluate(TensorDomain::noReductions(tv->getRFactorDomain())
-                          .back()
-                          ->extent())
-            .as<int64_t>();
-    supported_vec_size.epilogue =
-        std::min(supported_vec_size.epilogue, 16l / dataTypeSize(tv->dtype()));
-    supported_vec_size.epilogue = std::min(
-        supported_vec_size.epilogue, scheduler_utils::maxVectorizationWidth(N));
   }
-  // For INPUT_C role tensors, we do not necessarily know which axis we would
-  // like to vectorize, so we set vectorization based on dtype instead here
-  // until a more complete analysis is implemented.
-  if (const auto c_it = roles_map.find(MatmulRole::INPUT_C);
-      c_it != roles_map.end()) {
-    for (TensorView* tv : c_it->second) {
-      supported_vec_size.epilogue = std::min(
-          supported_vec_size.epilogue, 16l / dataTypeSize(tv->dtype()));
+
+  // Inspect the allocation domain of an operand input TensorView to determine
+  // vectorization width.
+  //
+  // We canonicalize dimensions by reordering them with the given ordering
+  // before merging all dimensions that have the same role. For a given operand,
+  // this might mean that the inner-most dimension gets reordered to be outer,
+  // even if it has the same role as the innermost dimension in the canonical
+  // ordering.
+  int64_t operandVectorization(TensorView* tv) {
+    // Check data pointer alignment
+    int64_t vec_size = ptrAndDTypeVec(tv);
+    if (vec_size == 1l) {
+      return vec_size;
     }
+
+    // Find the inner-most non-batch role for this tensor, and collect all
+    // ValGroups in that role, in the canonical ordering.
+    std::optional<MatmulDomain> vec_dim_role = std::nullopt;
+    for (int64_t i = (int64_t)(tv->getMaybeAllocationDomain().size()) - 1;
+         i >= 0;
+         --i) {
+      IterDomain* id = tv->getMaybeAllocationDomain()[i];
+      if (id->isDeviceDim() || id->isReduction() || id->isBroadcast()) {
+        continue;
+      }
+
+      ValGroup g = exact_graph_.toGroup(id);
+      MatmulDomain dim_role = dimRole(g);
+      if (dim_role == MatmulDomain::Batch) {
+        // We cannot vectorize in batch dimensions
+        break;
+      }
+      if (!vec_dim_role.has_value()) {
+        vec_dim_role = dim_role;
+      }
+    }
+    if (!vec_dim_role.has_value()) {
+      // Didn't find any dimensions to vectorize
+      return 1l;
+    }
+
+    // Extract dims with this role in the canonical ordering
+    std::vector<ValGroup> ordered_inner_dims;
+    for (const ValGroup& og : dim_ordering_) {
+      if (dimRole(og) == vec_dim_role.value()) {
+        ordered_inner_dims.push_back(og);
+      }
+    }
+
+    return std::min(vec_size, innerDimsVectorization(tv, ordered_inner_dims));
```

Comment:
Should we also check the strides of inner dim is 1, and the non-inner dim? Merged inner dims are considered as a single inner dim.

Reply:
Strides are checked in `innerDimsVectorization`. We actually only vectorize dims if the `TensorView` has those dimensions marked contiguous, but we do check that the strides of all those vectorized dims match the expected contiguous values. This includes the case of a single inner-most dim, in which case we check that the stride is 1: https://github.com/NVIDIA/Fuser/blob/6f9f79c68d931734485e1e873b2ee2b5c80dc552/csrc/scheduler/matmul_utils.cpp#L476-L484

Reply:
For the dims not contiguous, should we use its stride value to `std::min` the vectorization size? For example, what if a dimension has stride 1331? This case, we can not vectorize because the address will not always be aligned.

Reply:
Outer dims are handled just as you say later on in this function: https://github.com/NVIDIA/Fuser/blob/6f9f79c68d931734485e1e873b2ee2b5c80dc552/csrc/scheduler/matmul_utils.cpp#L503-L511